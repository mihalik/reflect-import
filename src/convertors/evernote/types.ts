import {ConversionError} from '../../types'

export class EvernoteConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'EvernoteConversionError'
  }
}
