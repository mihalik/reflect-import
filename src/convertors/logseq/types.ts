import {ConversionError} from '../../types'

export interface LogseqNote {
  'page-name': string
  id: string
  children: LogseqBlock[]
}

export interface LogseqBlock {
  format: string
  content: string
  properties: Record<string, string>
  id: string
  children: LogseqBlock[]
}

export class LogseqConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'LogseqConversionError'
  }
}
