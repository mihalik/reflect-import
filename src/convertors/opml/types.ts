import {ConversionError} from '../../types'

export class OpmlConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'OpmlConversionError'
  }
}
