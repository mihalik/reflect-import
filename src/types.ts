export const REFLECT_HOSTNAME = 'reflect.app'

export type ConvertedNote = {
  // A required html string
  html: string
  // An optional subject
  subject?: string
  // A list of note ids that are references from inside this note
  backlinkNoteIds?: string[]
  // The date the note was created
  createdAt?: number
  // The date the note was updated
  updatedAt?: number
}

export interface Convertor {
  convert(data: string): ConvertedNote
}

export interface ListConvertor {
  convert(data: string): ConvertedNote[]
}
export class ConversionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ConversionError'
  }
}
