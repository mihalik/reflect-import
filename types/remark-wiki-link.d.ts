declare module 'remark-wiki-link' {
  export default wikiLinkPlugin
}
