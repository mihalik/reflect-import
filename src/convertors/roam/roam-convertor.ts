import {DOM, domArrayToHtml, domToHtml} from '../../helpers/dom'
import {header1, list, listItem, taskListItem} from '../../helpers/generators'
import {ConvertOptions, Convertor, ConvertResponse, REFLECT_HOSTNAME} from '../../types'
import {markdownToHtml} from '../../helpers/markdown'
import {RoamConversionError, RoamConvertedNote, RoamNote, RoamNoteString} from './types'
import {parseDateFromSubject, parseNoteIdFromSubject, validateTime} from './roam-helpers'
import {first} from 'lodash'

export class RoamConvertor implements Convertor {
  graphId: string
  linkHost: string

  constructor({
    graphId,
    linkHost = REFLECT_HOSTNAME,
  }: {
    graphId: string
    linkHost?: string
  }) {
    this.graphId = graphId
    this.linkHost = linkHost
  }

  accept = {'application/json': ['.json']}

  convert({data}: ConvertOptions): ConvertResponse {
    const roamNotes = JSON.parse(data) as RoamNote[]

    if (!Array.isArray(roamNotes)) {
      throw new RoamConversionError('Roam export must be an array of notes')
    }

    const notes = roamNotes.map((note) => this.convertRoamNote(note))

    return {notes}
  }

  private convertRoamNote(note: RoamNote): RoamConvertedNote {
    const {html, backlinkNoteIds} = this.extractHtmlAndBacklinks(note)

    const updated = note['edit-time']
    const minChildCreated = first(
      (note.children ?? []).map((child) => child['create-time']).sort(),
    )
    const titleDate = parseDateFromSubject(note.title)

    return {
      id: note.uid,
      html,
      subject: note.title,
      backlinkNoteIds,
      isDaily: !!titleDate,
      createdAt: validateTime(titleDate?.getTime() ?? minChildCreated),
      updatedAt: validateTime(updated),
    }
  }

  private extractHtmlAndBacklinks(note: RoamNote) {
    let dom = list()
    const aggregBacklinkNoteIds = new Set<string>()

    if (note.children) {
      const listItems: DOM[] = []

      for (const child of note.children) {
        const {html, backlinkNoteIds} = this.parseListItem(child)

        listItems.push(html)
        backlinkNoteIds.forEach((id) => aggregBacklinkNoteIds.add(id))
      }

      dom = list(listItems)
    }

    return {
      html: domArrayToHtml([header1(note.title), dom]),
      backlinkNoteIds: Array.from(aggregBacklinkNoteIds),
    }
  }

  private parseListItem(noteString: RoamNoteString, aggregNoteIds = new Set<string>()) {
    let string = this.convertRoamTagsToBacklinks(noteString.string?.trim() ?? '')

    let taskChecked: boolean | undefined

    string = string.replace(LINKED_TODO_UNFINISHED_STRING, TODO_UNFINISHED_STRING)
    string = string.replace(LINKED_TODO_FINISHED_STRING, TODO_FINISHED_STRING)

    if (string.startsWith(TODO_UNFINISHED_STRING)) {
      string = string.replace(TODO_UNFINISHED_STRING, '')
      taskChecked = false
    } else if (string.startsWith(TODO_FINISHED_STRING)) {
      string = string.replace(TODO_FINISHED_STRING, '')
      taskChecked = true
    }

    const {html: itemContent, backlinkNoteIds} = markdownToHtml(string, {
      graphId: this.graphId,
      linkHost: this.linkHost,
      constructsToDisable: ['thematicBreak', 'list', 'headingAtx'],
      pageResolver: parseNoteIdFromSubject,
    })

    backlinkNoteIds.forEach((id) => aggregNoteIds.add(id))

    let itemChildren: DOM = ''

    if (noteString.children) {
      const listItems = noteString.children.map((child) => {
        const {html} = this.parseListItem(child, aggregNoteIds)
        return html
      })
      itemChildren = list(listItems)
    }

    const dom =
      taskChecked === undefined
        ? listItem(domArrayToHtml([itemContent, itemChildren]))
        : taskListItem(domArrayToHtml([itemContent, itemChildren]), {
            checked: taskChecked,
          })

    return {
      html: domToHtml(dom),
      backlinkNoteIds: Array.from(aggregNoteIds),
    }
  }

  private convertRoamTagsToBacklinks(str: string) {
    return str.replace(/(^|\s)#\[\[/g, '$1[[').replace(/(^|\s)#([\w-]+)/g, '$1[[$2]]')
  }
}

const LINKED_TODO_UNFINISHED_STRING = '{{[[TODO]]}}'
const LINKED_TODO_FINISHED_STRING = '{{[[DONE]]}}'

const TODO_UNFINISHED_STRING = '{{TODO}}'
const TODO_FINISHED_STRING = '{{DONE}}'
