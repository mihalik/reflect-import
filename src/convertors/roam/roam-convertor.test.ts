import {RoamConvertor} from './roam-convertor'
import {RoamNote, RoamNoteString} from './types'

const ROAM_SAMPLE: RoamNote = {
  title: 'The Great CEO Within',
  uid: '09-14-2020',
  'edit-time': 1600057101270,
  children: [
    {
      string: 'Author:: #[[Matt Mochary]]',
      'create-time': 1599161754087,
      uid: '-rC8kxJH4',
      'edit-time': 1600958531049,
    },
    {
      string: '**Chapter 1: Getting Started**',
      'create-time': 1576448157574,
      children: [
        {
          string:
            'This is explained clearly and thoroughly in Disciplined Entrepreneurship by Bill Aulet. I won’t repeat or even summarize what he wrote. If you haven’t yet launched or achieved more than $1 million of revenue, go read Bill’s book first.',
          'create-time': 1576093423770,
          uid: 'QhiYH-2V4',
          'edit-time': 1600716930617,
        },
      ],
      uid: 'bn1X-uy0x',
      'edit-time': 1600716930617,
    },
  ],
}

const CONTENT_SAMPLE = {
  type: 'doc',
  content: [
    {
      type: 'heading',
      attrs: {level: 1},
      content: [{type: 'text', text: 'The Great CEO Within'}],
    },
    {
      type: 'bulletList',
      content: [
        {
          type: 'listItem',
          attrs: {closed: false, nested: false},
          content: [
            {
              type: 'paragraph',
              content: [
                {type: 'text', text: 'Author:: '},
                {
                  attrs: {
                    id: 'mattmochary',
                    label: 'Matt Mochary',
                  },
                  type: 'backlink',
                },
              ],
            },
          ],
        },
        {
          type: 'listItem',
          attrs: {closed: false, nested: false},
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  type: 'text',
                  marks: [{type: 'bold'}],
                  text: 'Chapter 1: Getting Started',
                },
              ],
            },
            {
              type: 'bulletList',
              content: [
                {
                  type: 'listItem',
                  attrs: {
                    closed: false,
                    nested: false,
                  },
                  content: [
                    {
                      type: 'paragraph',
                      content: [
                        {
                          type: 'text',
                          text:
                            'This is explained clearly and thoroughly in Disciplined Entrepreneurship by Bill Aulet. I won’t repeat or even summarize what he wrote. If you haven’t yet launched or achieved more than $1 million of revenue, go read Bill’s book first.',
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}

const htmlFromRoamNote = (note: RoamNote, graphId: string) =>
  new RoamConvertor({graphId}).toHtml(JSON.stringify(note))

test('parseContentFromHTML', () => {
  const result = htmlFromRoamNote(ROAM_SAMPLE, '123')

  expect(result).toMatchSnapshot()
})

describe('generateContentFromRoamNote', () => {
  it('handles both formats of todos', () => {
    const noteString: RoamNoteString = {
      string: 'parent',
      'create-time': 12345,
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: `{{TODO}}  todo 1`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
        {
          string: `{{[[TODO]]}}  todo 2`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
      ],
    }

    const note: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [noteString],
    }

    const doc = htmlFromRoamNote(note, '123')
    expect(doc).toMatchSnapshot()
  })

  it('handles list with mixed bullets and tasks', () => {
    const noteString: RoamNoteString = {
      string: 'parent',
      'create-time': 12345,
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: `{{TODO}}  todo 1`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
        {
          string: `{{DONE}}  todo 2`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
        {
          string: `bullet 1`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
        {
          string: `bullet 2`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
        {
          string: `{{TODO}}  todo 3`,
          'create-time': 12345,
          uid: '456',
          'edit-time': 12345,
          children: [],
        },
      ],
    }

    const note: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [noteString],
    }

    const doc = htmlFromRoamNote(note, '123')

    expect(doc).toMatchSnapshot()
  })
})

// it('handles code blocks', () => {
//   const node = generateListItemFromRoamNoteString(
//     {
//       string: '```javascript\nlonger code snippet\n```',
//       'create-time': 12345,
//       uid: '123',
//       'edit-time': 12345,
//       children: [],
//     },
//     '123',
//   )

//   const html = htmlFromNode(node)

//   expect(html).toEqual(
//     `<li class="remirror-list-item-with-custom-mark"><span class="remirror-list-item-marker-container"><div class="remirror-collapsible-list-item-button"></div></span><div><pre spellcheck="false" class="language-javascript"><code data-code-block-language="javascript">longer code snippet\n</code></pre></div></li>`,
//   )
// })

//   it('handles ordered lists', () => {
//     const node = generateListItemFromRoamNoteString(
//       {
//         string: '1. foo',
//         'create-time': 12345,
//         uid: '123',
//         'edit-time': 12345,
//         children: [],
//       },
//       '123',
//     )

//     const html = htmlFromNode(node)
//     expect(html).toEqual(
//       `<li class="remirror-list-item-with-custom-mark"><span class="remirror-list-item-marker-container"><div class="remirror-collapsible-list-item-button"></div></span><div><p>1. foo</p></div></li>`,
//     )
//   })

//   it('handles blockquotes', () => {
//     const node = generateListItemFromRoamNoteString(
//       {
//         string: '> hello world',
//         'create-time': 12345,
//         uid: '123',
//         'edit-time': 12345,
//         children: [],
//       },
//       '123',
//     )

//     const html = htmlFromNode(node)
//     expect(formatHtml(html)).toMatchInlineSnapshot(`
//       "<li class=\\"remirror-list-item-with-custom-mark\\">
//         <span class=\\"remirror-list-item-marker-container\\"
//           ><div class=\\"remirror-collapsible-list-item-button\\"></div
//         ></span>
//         <div>
//           <blockquote><p>hello world</p></blockquote>
//         </div>
//       </li>
//       "
//     `)
//   })

//   it('generates content with the roam note', () => {
//     const noteString: RoamNoteString = {
//       string: 'hello world',
//       'create-time': 12345,
//       uid: '123',
//       'edit-time': 12345,
//       children: [
//         {
//           string: `everything is **awesome** #foo`,
//           'create-time': 12345,
//           uid: '456',
//           'edit-time': 12345,
//           children: [
//             {
//               string: `wahooo!\n\ngood times`,
//               'create-time': 12345,
//               uid: '789',
//               'edit-time': 12345,
//             },
//           ],
//         },
//       ],
//     }

//     const note: RoamNote = {
//       title: 'note1',
//       uid: '123',
//       'edit-time': 12345,
//       children: [noteString],
//     }

//     const doc = htmlFromRoamNote(note, '123')
//     const html = htmlFromNode(doc)

//     expect(formatHtml(html)).toMatchInlineSnapshot(`
//       "<h1>note1</h1>
//       <ul>
//         <li class=\\"remirror-list-item-with-custom-mark\\">
//           <span class=\\"remirror-list-item-marker-container\\"
//             ><div class=\\"remirror-collapsible-list-item-button\\"></div
//           ></span>
//           <div>
//             <p>hello world</p>
//             <ul>
//               <li class=\\"remirror-list-item-with-custom-mark\\">
//                 <span class=\\"remirror-list-item-marker-container\\"
//                   ><div class=\\"remirror-collapsible-list-item-button\\"></div
//                 ></span>
//                 <div>
//                   <p>
//                     everything is <strong>awesome</strong>
//                     <a href=\\"https://reflect.app/g/123/foo\\">foo</a>
//                   </p>
//                   <ul>
//                     <li class=\\"remirror-list-item-with-custom-mark\\">
//                       <span class=\\"remirror-list-item-marker-container\\"
//                         ><div class=\\"remirror-collapsible-list-item-button\\"></div
//                       ></span>
//                       <div>
//                         <p>wahooo!</p>
//                         <p>good times</p>
//                       </div>
//                     </li>
//                   </ul>
//                 </div>
//               </li>
//             </ul>
//           </div>
//         </li>
//       </ul>
//       "
//     `)
//   })

//   it('handles backlinks', () => {
//     const noteString: RoamNoteString = {
//       string: 'link to [[another note]]',
//       'create-time': 12345,
//       uid: '123',
//       'edit-time': 12345,
//       children: [],
//     }

//     const note: RoamNote = {
//       title: 'note1',
//       uid: '123',
//       'edit-time': 12345,
//       children: [noteString],
//     }

//     const doc = htmlFromRoamNote(note, '123')

//     const backlink = doc.content
//       .child(1)
//       .content.child(0)
//       .content.child(0)
//       .content.child(1)

//     expect(backlink.attrs['id']).toEqual('anothernote')
//   })
// })

// describe('generateListFromRoamNoteString', () => {
//   it('doesnt fail with empty string', () => {
//     const note: RoamNoteString = {
//       string: '',
//       'create-time': 12345,
//       uid: '123',
//       'edit-time': 12345,
//     }

//     const nodes = collectChildren(generateListItemFromRoamNoteString(note, '123'))
//     const doc = g.doc([g.header('note'), ...nodes])
//     const html = htmlFromNode(doc)

//     expect(html).toEqual('<h1>note</h1><p></p>')
//   })
// })
