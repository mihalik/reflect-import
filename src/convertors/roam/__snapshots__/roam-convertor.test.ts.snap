// Vitest Snapshot v1

exports[`RoamConvertor > generateContentFromRoamNote > handles both formats of todos 1`] = `"<h1>note1</h1><ul><li><p>parent</p><ul><li><input type=\\"checkbox\\" ><p>todo 1</p><ul></ul></li><li><input type=\\"checkbox\\" ><p>todo 2</p><ul></ul></li></ul></li></ul>"`;

exports[`RoamConvertor > generateContentFromRoamNote > handles list with mixed bullets and tasks 1`] = `"<h1>note1</h1><ul><li><p>parent</p><ul><li><input type=\\"checkbox\\" ><p>todo 1</p><ul></ul></li><li><input type=\\"checkbox\\" checked><p>todo 2</p><ul></ul></li><li><p>bullet 1</p><ul></ul></li><li><p>bullet 2</p><ul></ul></li><li><input type=\\"checkbox\\" ><p>todo 3</p><ul></ul></li></ul></li></ul>"`;
